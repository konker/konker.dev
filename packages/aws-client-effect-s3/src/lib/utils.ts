/* eslint-disable fp/no-unused-expression,fp/no-let,fp/no-nil */
import type { Readable } from 'node:stream';

import { toError } from '@konker.dev/tiny-error-fp/lib';
import * as Effect from 'effect/Effect';

import type { PromiseDependentWritableStream } from './PromiseDependentWritableStream';

/**
 * Wait for a readable stream to fully pipe to a PromiseDependentWritableStream sink
 */
export function waitForPromiseDependentWritableStreamPipe(
  readStream: Readable,
  writeStream: PromiseDependentWritableStream
): Effect.Effect<number, Error> {
  return Effect.tryPromise({
    try: () =>
      new Promise((resolve, reject) => {
        let size = 0;
        readStream.on('data', (data: string) => {
          // eslint-disable-next-line fp/no-mutation
          size = size + data.length;
        });
        readStream.on('error', reject);
        readStream.pipe(writeStream);
        readStream.resume();
        if (writeStream.promise) writeStream.promise.then(() => resolve(size)).catch(reject);
        else reject(Error('waitForPromiseDependentWritableStreamPipe called without a stream promise'));
      }),
    catch: toError,
  });
}
