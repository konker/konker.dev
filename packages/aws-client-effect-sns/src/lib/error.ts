export const TAG = 'SnsError';

export type SnsError = {
  readonly _tag: typeof TAG;
  readonly _Params: unknown;
  readonly message: string;
  readonly cause: unknown;
};

export const toSnsError =
  <I>(params: I) =>
  (x: unknown): SnsError => {
    return {
      _tag: TAG,
      _Params: params,
      message: typeof x === 'object' && x && 'message' in x ? (x as any).message : String(x),
      cause: x,
    };
  };
