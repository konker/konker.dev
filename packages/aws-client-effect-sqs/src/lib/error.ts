export const TAG = 'SqsError';

export type SqsError = {
  readonly _tag: typeof TAG;
  readonly _Params: unknown;
  readonly message: string;
  readonly cause: unknown;
};

export const toSqsError =
  <I>(params: I) =>
  (x: unknown): SqsError => {
    return {
      _tag: TAG,
      _Params: params,
      message: typeof x === 'object' && x && 'message' in x ? (x as any).message : String(x),
      cause: x,
    };
  };
