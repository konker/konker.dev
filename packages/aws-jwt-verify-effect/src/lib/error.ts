export const TAG = 'JwtVerifyError';

export type JwtVerifyError = {
  readonly _tag: typeof TAG;
  readonly message: string;
  readonly cause: unknown;
};

export const toJwtVerifyError = (x: unknown): JwtVerifyError => {
  return {
    _tag: TAG,
    message: typeof x === 'object' && x && 'message' in x ? (x as any).message : String(x),
    cause: x,
  };
};
