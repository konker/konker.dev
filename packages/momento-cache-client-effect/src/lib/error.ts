import type { TinyError } from '@konker.dev/tiny-error-fp';
import { toTinyError } from '@konker.dev/tiny-error-fp';

export const ERROR_TAG = 'MomentoClientError' as const;
export type ERROR_TAG = typeof ERROR_TAG;

export type MomentoClientError = TinyError<ERROR_TAG>;
export const toMomentoClientError = toTinyError<ERROR_TAG>(ERROR_TAG);
