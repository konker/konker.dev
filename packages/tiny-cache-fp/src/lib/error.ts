import type { TinyError } from '@konker.dev/tiny-error-fp';
import { toTinyError } from '@konker.dev/tiny-error-fp';

export const ERROR_TAG = 'CacheError' as const;
export type ERROR_TAG = typeof ERROR_TAG;

export type CacheError = TinyError<ERROR_TAG>;
export const toCacheError = toTinyError<ERROR_TAG>(ERROR_TAG);
