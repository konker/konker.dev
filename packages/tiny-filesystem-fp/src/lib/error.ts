export const TAG = 'TinyFileSystemError';

export type TinyFileSystemError = {
  readonly _tag: typeof TAG;
  readonly message: string;
  readonly cause: unknown;
};

export function toTinyFileSystemError(x: unknown): TinyFileSystemError {
  return {
    _tag: TAG,
    message: typeof x === 'object' && x && 'message' in x ? (x as any).message : String(x),
    cause: x,
  };
}
