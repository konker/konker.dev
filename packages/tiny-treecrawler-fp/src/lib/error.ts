import { toTinyError } from '@konker.dev/tiny-error-fp';

export const ERROR_TAG = 'TinyTreeCrawlerError';
export type ERROR_TAG = typeof ERROR_TAG;

export const toTinyTreeCrawlerError = toTinyError<ERROR_TAG>(ERROR_TAG);
export type TinyTreeCrawlerError = ReturnType<typeof toTinyTreeCrawlerError>;
