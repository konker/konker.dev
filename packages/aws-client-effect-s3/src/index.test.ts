import type {
  DeleteObjectCommandInput,
  GetObjectCommandInput,
  HeadObjectCommandInput,
  ListObjectsV2CommandInput,
  PutObjectCommandInput,
} from '@aws-sdk/client-s3';
import * as s3Client from '@aws-sdk/client-s3';
import { S3Client } from '@aws-sdk/client-s3';
import * as P from '@konker.dev/effect-ts-prelude';
import { mockClient } from 'aws-sdk-client-mock';
import { beforeAll, beforeEach, describe, expect, it } from 'vitest';

import * as unit from './index';
import { S3ClientDeps } from './index';
import { TAG } from './lib/error';

const s3Mock = mockClient(s3Client.S3Client);

describe('aws-client-effect-s3', () => {
  // eslint-disable-next-line fp/no-let
  let deps: unit.S3ClientDeps;

  beforeAll(() => {
    // eslint-disable-next-line fp/no-mutation
    deps = unit.S3ClientDeps.of({
      s3Client: new s3Client.S3Client({}),
    });
  });

  describe('Error tag', () => {
    it('should export as expected', () => {
      expect(unit.S3_ERROR_TAG).toEqual('@konker.dev/aws-client-effect-s3/S3Error');
    });
  });

  // ------------------------------------------------------------------------
  describe('defaultS3ClientFactory', () => {
    it('should work as expected', async () => {
      expect(unit.defaultS3ClientFactory({})).toBeInstanceOf(S3Client);
    });
  });

  // ------------------------------------------------------------------------
  describe('defaultS3ClientFactoryDeps', () => {
    it('should work as expected', async () => {
      const actualEffect = P.pipe(
        unit.S3ClientFactoryDeps,
        P.Effect.map((deps) => deps.s3ClientFactory),
        unit.defaultS3ClientFactoryDeps
      );
      const actual = P.Effect.runSync(actualEffect);
      expect(actual).toBeInstanceOf(Function);
    });
  });

  // ------------------------------------------------------------------------
  describe('defaultS3ClientDeps', () => {
    it('should work as expected', async () => {
      const actualEffect = P.pipe(
        unit.S3ClientDeps,
        P.Effect.map((deps) => deps.s3Client),
        unit.defaultS3ClientDeps({})
      );
      const actual = P.Effect.runSync(actualEffect);
      expect(actual).toBeInstanceOf(S3Client);
    });
  });

  // ------------------------------------------------------------------------
  describe('GetObjectCommand', () => {
    beforeEach(() => {
      s3Mock.reset();
    });

    it('should work as expected', async () => {
      s3Mock.on(s3Client.GetObjectCommand).resolves({ Body: 'Hello, World!' } as any);

      const params: GetObjectCommandInput = { Bucket: 'test-bucket', Key: 'test-key' };
      const expected = { Body: 'Hello, World!', _Params: params };
      const command = P.pipe(unit.GetObjectCommandEffect(params), P.Effect.provideService(S3ClientDeps, deps));
      await expect(P.Effect.runPromise(command)).resolves.toStrictEqual(expected);
      expect(s3Mock.calls().length).toEqual(1);
    });

    it('should fail as expected', async () => {
      s3Mock.on(s3Client.GetObjectCommand).rejects({ $metadata: { httpStatusCode: 404 }, code: 'NotFound' } as any);

      const params: GetObjectCommandInput = { Bucket: 'test-bucket', Key: 'test-key-not-found' };
      const command = P.pipe(unit.GetObjectCommandEffect(params), P.Effect.provideService(S3ClientDeps, deps));
      await expect(P.Effect.runPromise(command)).rejects.toThrowErrorMatchingSnapshot(
        JSON.stringify({ _tag: TAG, message: 'NotFound' })
      );
      expect(s3Mock.calls().length).toEqual(1);
    });
  });

  // ------------------------------------------------------------------------
  describe('PutObjectCommand', () => {
    beforeEach(() => {
      s3Mock.reset();
    });

    it('should work as expected', async () => {
      s3Mock.on(s3Client.PutObjectCommand).resolves({} as any);

      const params: PutObjectCommandInput = { Bucket: 'test-bucket', Key: 'test-key', Body: 'test-body' };
      const expected = { _Params: params };
      const command = P.pipe(unit.PutObjectCommandEffect(params), P.Effect.provideService(S3ClientDeps, deps));
      await expect(P.Effect.runPromise(command)).resolves.toStrictEqual(expected);
      expect(s3Mock.calls().length).toEqual(1);
    });

    it('should fail as expected', async () => {
      s3Mock.on(s3Client.PutObjectCommand).rejects({ $metadata: { httpStatusCode: 404 }, code: 'NotFound' } as any);

      const params: PutObjectCommandInput = { Bucket: 'test-bucket-not-found', Key: 'test-key' };
      const command = P.pipe(unit.PutObjectCommandEffect(params), P.Effect.provideService(S3ClientDeps, deps));
      await expect(P.Effect.runPromise(command)).rejects.toThrowErrorMatchingSnapshot(
        JSON.stringify({ _tag: TAG, message: 'NotFound' })
      );
      expect(s3Mock.calls().length).toEqual(1);
    });
  });

  // ------------------------------------------------------------------------
  describe('HeadObjectCommand', () => {
    beforeEach(() => {
      s3Mock.reset();
    });

    it('should work as expected', async () => {
      s3Mock.on(s3Client.HeadObjectCommand).resolves({ Body: 'Hello, World!' } as any);

      const params: HeadObjectCommandInput = { Bucket: 'test-bucket', Key: 'test-key' };
      const expected = { Body: 'Hello, World!', _Params: params };
      const command = P.pipe(unit.HeadObjectCommandEffect(params), P.Effect.provideService(S3ClientDeps, deps));
      await expect(P.Effect.runPromise(command)).resolves.toStrictEqual(expected);
      expect(s3Mock.calls().length).toEqual(1);
    });

    it('should fail as expected', async () => {
      s3Mock.on(s3Client.HeadObjectCommand).rejects({ $metadata: { httpStatusCode: 404 }, code: 'NotFound' } as any);

      const params: HeadObjectCommandInput = { Bucket: 'test-bucket', Key: 'test-key-not-found' };
      const command = P.pipe(unit.HeadObjectCommandEffect(params), P.Effect.provideService(S3ClientDeps, deps));
      await expect(P.Effect.runPromise(command)).rejects.toThrowErrorMatchingSnapshot(
        JSON.stringify({ _tag: TAG, message: 'NotFound' })
      );
      expect(s3Mock.calls().length).toEqual(1);
    });
  });

  // ------------------------------------------------------------------------
  describe('DeleteObjectCommand', () => {
    beforeEach(() => {
      s3Mock.reset();
    });

    it('should work as expected', async () => {
      s3Mock.on(s3Client.DeleteObjectCommand).resolves({} as any);

      const params: DeleteObjectCommandInput = { Bucket: 'test-bucket', Key: 'test-key' };
      const expected = { _Params: params };
      const command = P.pipe(unit.DeleteObjectCommandEffect(params), P.Effect.provideService(S3ClientDeps, deps));
      await expect(P.Effect.runPromise(command)).resolves.toStrictEqual(expected);
      expect(s3Mock.calls().length).toEqual(1);
    });

    it('should fail as expected', async () => {
      s3Mock.on(s3Client.DeleteObjectCommand).rejects({ $metadata: { httpStatusCode: 404 }, code: 'NotFound' } as any);

      const params: DeleteObjectCommandInput = { Bucket: 'test-bucket', Key: 'test-key-not-found' };
      const command = P.pipe(unit.DeleteObjectCommandEffect(params), P.Effect.provideService(S3ClientDeps, deps));
      await expect(P.Effect.runPromise(command)).rejects.toThrowErrorMatchingSnapshot(
        JSON.stringify({ _tag: TAG, message: 'NotFound' })
      );
      expect(s3Mock.calls().length).toEqual(1);
    });
  });

  // ------------------------------------------------------------------------
  describe('ListObjectsV2Command', () => {
    beforeEach(() => {
      s3Mock.reset();
    });

    it('should work as expected', async () => {
      s3Mock.on(s3Client.ListObjectsV2Command).resolves({ Contents: [] });

      const params: ListObjectsV2CommandInput = { Bucket: 'test-bucket' };
      const expected = { Contents: [], _Params: params };
      const command = P.pipe(unit.ListObjectsV2CommandEffect(params), P.Effect.provideService(S3ClientDeps, deps));
      await expect(P.Effect.runPromise(command)).resolves.toStrictEqual(expected);
      expect(s3Mock.calls().length).toEqual(1);
    });

    it('should fail as expected', async () => {
      s3Mock.on(s3Client.ListObjectsV2Command).rejects({ $metadata: { httpStatusCode: 404 }, code: 'NotFound' } as any);

      const params: ListObjectsV2CommandInput = { Bucket: 'test-bucket-not-found' };
      const command = P.pipe(unit.ListObjectsV2CommandEffect(params), P.Effect.provideService(S3ClientDeps, deps));
      await expect(P.Effect.runPromise(command)).rejects.toThrowErrorMatchingSnapshot(
        JSON.stringify({ _tag: TAG, message: 'NotFound' })
      );
      expect(s3Mock.calls().length).toEqual(1);
    });
  });
});
