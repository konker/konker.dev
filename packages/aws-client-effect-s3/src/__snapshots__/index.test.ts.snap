// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`aws-client-effect-s3 DeleteObjectCommand should fail as expected: {"_tag":"@konker.dev/aws-client-effect-s3/S3Error","message":"NotFound"} 1`] = `"{"_tag":"@konker.dev/aws-client-effect-s3/S3Error","_Params":{"Bucket":"test-bucket","Key":"test-key-not-found"},"message":"NotFound","cause":{"$metadata":{"httpStatusCode":404},"code":"NotFound"}}"`;

exports[`aws-client-effect-s3 GetObjectCommand should fail as expected: {"_tag":"@konker.dev/aws-client-effect-s3/S3Error","message":"NotFound"} 1`] = `"{"_tag":"@konker.dev/aws-client-effect-s3/S3Error","_Params":{"Bucket":"test-bucket","Key":"test-key-not-found"},"message":"NotFound","cause":{"$metadata":{"httpStatusCode":404},"code":"NotFound"}}"`;

exports[`aws-client-effect-s3 HeadObjectCommand should fail as expected: {"_tag":"@konker.dev/aws-client-effect-s3/S3Error","message":"NotFound"} 1`] = `"{"_tag":"@konker.dev/aws-client-effect-s3/S3Error","_Params":{"Bucket":"test-bucket","Key":"test-key-not-found"},"message":"NotFound","cause":{"$metadata":{"httpStatusCode":404},"code":"NotFound"}}"`;

exports[`aws-client-effect-s3 ListObjectsV2Command should fail as expected: {"_tag":"@konker.dev/aws-client-effect-s3/S3Error","message":"NotFound"} 1`] = `"{"_tag":"@konker.dev/aws-client-effect-s3/S3Error","_Params":{"Bucket":"test-bucket-not-found"},"message":"NotFound","cause":{"$metadata":{"httpStatusCode":404},"code":"NotFound"}}"`;

exports[`aws-client-effect-s3 PutObjectCommand should fail as expected: {"_tag":"@konker.dev/aws-client-effect-s3/S3Error","message":"NotFound"} 1`] = `"{"_tag":"@konker.dev/aws-client-effect-s3/S3Error","_Params":{"Bucket":"test-bucket-not-found","Key":"test-key"},"message":"NotFound","cause":{"$metadata":{"httpStatusCode":404},"code":"NotFound"}}"`;
